### PR TITLE
Lazy init order in set device, should not be called in getDevCount

### PR DESCRIPTION
**Pull request by csarofeen**

### CODE REVIEW

**Comment by csarofeen** [comment]
https://github.com/pytorch/pytorch/issues/4903

File: torch/cuda/__init__.py [hunk 1]
```diff
@@ -308,7 +308,6 @@ def stream(stream):
 def device_count():
     """Returns the number of GPUs available."""
     if is_available():
-        _lazy_init()
```
Comment (apaszke) [inline torch/cuda/__init__.py]: Won't `is_available()` initialize CUDA anyway?

File: torch/cuda/__init__.py [hunk 2]
```diff
@@ -215,10 +215,10 @@ def __init__(self, idx):
     def __enter__(self):
         if self.idx is -1:
             return
-        _lazy_init()
         self.prev_idx = torch._C._cuda_getDevice()
         if self.prev_idx != self.idx:
             torch._C._cuda_setDevice(self.idx)
+        _lazy_init()
```
Comment (apaszke) [inline torch/cuda/__init__.py]: Does `cudaGetDevice` initialize a context?

**Review by apaszke — CHANGES_REQUESTED** [review]
Can you please revert the pybind submodule update?

**Comment by csarofeen** [comment]
The biggest issue seems to be _lazy_init as it is allocating a good chunk of memory that can end up on the wrong device fairly easily the way things are now. Calling device count in pytorch was allocating about 1.5GB on my first device. Device count in pure CUDA was creating ~10MB on each device.

File: torch/cuda/__init__.py (hunk 2 above)
Comment (albanD) [reply torch/cuda/__init__.py]: From my tests using `torch.cuda.set_device(xx)`, no.

File: torch/cuda/__init__.py (hunk 1 above)
Comment (albanD) [reply torch/cuda/__init__.py]: As of the current implementation no. And `torch._C._cuda_getDeviceCount()` does not require it to be initialized (again from tests from python).

**Pull request merged**